<a id='top'></a>

# CSCI 3022: Intro to Data Science - Spring 2019 Practicum 2
***

This practicum is due on Canvas by **11:59 PM on Friday May 3**. Your solutions to theoretical questions should be done in Markdown/MathJax directly below the associated question.  Your solutions to computational questions should include any specified Python code and results as well as written commentary on your conclusions.  

**Here are the rules:** 

1. All work, code and analysis, must be your own. 
1. You may use your course notes, posted lecture slides, textbooks, in-class notebooks, and homework solutions as resources.  You may also search online for answers to general knowledge questions like the form of a probability distribution function or how to perform a particular operation in Python/Pandas. 
1. This is meant to be like a coding portion of your midterm exam. So, the instructional team will be much less helpful than we typically are with homework. For example, we will not check answers, help debug your code, and so on.
1. If something is left open-ended, it is because we want to see how you approach the kinds of problems you will encounter in the wild, where it will not always be clear what sort of tests/methods should be applied. Feel free to ask clarifying questions though.
2. You may **NOT** post to message boards or other online resources asking for help.
3. You may **NOT** copy-paste solutions *from anywhere*.
4. You may **NOT** collaborate with classmates or anyone else.
5. In short, **your work must be your own**. It really is that simple.

Violation of the above rules will result in an immediate academic sanction (*at the very least*, you will receive a 0 on this practicum or an F in the course, depending on severity), and a trip to the Honor Code Council.

**By submitting this assignment, you agree to abide by the rules given above.**

***

**Name**:  

Tyler Valentine


**NOTES**: 

- You may not use late days on the practicums nor can you drop your practicum grades. 
- If you have a question for us, post it as a **PRIVATE** message on Piazza.  If we decide that the question is appropriate for the entire class, then we will add it to a Practicum clarifications thread. 
- Do **NOT** load or use any Python packages that are not available in Anaconda 3.6. 
- Some problems with code may be autograded.  If we provide a function API **do not** change it.  If we do not provide a function API then you're free to structure your code however you like. 
- Submit only this Jupyter notebook to Canvas.  Do not compress it using tar, rar, zip, etc. 
- This should go without saying, but... For any question that asks you to calculate something, you **must show all work to receive credit**. Sparse or nonexistent work will receive sparse or nonexistent credit.

---
**Shortcuts:**  [Problem 1](#p1) | [Problem 2](#p2) | [Bottom](#bot)

---

In [76]:
import numpy as np 
from scipy import stats
import statsmodels.api as sm 
import pandas as pd
import matplotlib.pylab as plt
%matplotlib inline
import statsmodels.api as sm 

<br>

---
<a id='p1'></a>
[Back to top](#top)

### [30 points] Problem 1:  What's the best Oreo flavor?

> <img src="https://media2.s-nbcnews.com/j/newscms/2015_52/1351976/151223-watermelon-oreos-flickr-yh-1234p_a4c4ba1016261e5d799a789666948a6e.fit-760w.jpg" style="width: 400px;"/>
**Figure 1.** Watermelon Oreos. These are undeniably the absolute _worst_ Oreo flavor, but not a part of our data set. If they were, the problem would be too easy.

You're skipping down the road one day, singing a merry tune without a care in the world. Every so often you hop and click your heels together. After one such hop, though, you stumble and nearly trip. Under your feet is a fun data set!

In particular, this rascally data set includes ratings from 5 different individuals for 12 different types of Oreo cookie. Because some of the raters are allergic to disgusting flavors of Oreo, not every taster tasted every type of cookie. More information about the data set can be found [here](https://www.kaggle.com/rtatman/oreo-flavors-tastetest-ratings).

As you examine the data, a booming voice from the sky instructs you to analyze the data to answer two key questions:
1. is there some difference in mean rating given to different flavors of Oreo? and
2. which flavor (out of those in the data set) is the best?

**Part A:** Load the data set as a Pandas DataFrame. Then, process it as you see fit. For the analysis, you might create a new DataFrame, or other kinds of Python objects to use the data. Do what feels right, and include an **explanation** in Markdown of how you are processing the data. You **may not** alter the original data file.

In [77]:
dfOreo = pd.read_csv("D:\Valentine\CU\CSCI-3022\Practicum_2\oreo_rankings.csv", na_values = '-')
dfOreo2 = pd.read_csv("D:\Valentine\CU\CSCI-3022\Practicum_2\oreo_rankings.csv", na_values = '-')
#dfOreo.dropna()
del (dfOreo['notes_and_discussion'])
del (dfOreo['oreo_flavor'])
dfOreo.head(12)

rater_n  rater_c  rater_r  rater_k  rater_a  average_across_raters
0       1.5      4.0      2.0      NaN      NaN                 2.5000
1       2.5      3.0      4.0      NaN     1.50                 2.7500
2       4.5      3.5      1.5      NaN     3.50                 3.2500
3       3.5      3.0      3.5      3.0     4.00                 3.4000
4       2.5      5.0      2.5      4.0     3.00                 3.4000
5       3.5      4.0      3.0      3.0     3.50                 3.4000
6       4.0      3.0      4.5      4.0     2.50                 3.6000
7       2.5      1.0      5.0      5.0     5.00                 3.7000
8       3.0      4.0      4.0      NaN     4.25                 3.8125
9       3.5      4.5      3.0      4.5     4.50                 4.0000
10      3.5      4.0      4.5      4.0     4.25                 4.0500
11      NaN      4.5      4.5      4.5     4.00                 4.3750

**ANSWER PART A**

In order for the data to be more useable for later parts of this question I got rid of all of of the columns that didn't contain numerical data. I also set data points that had no values to NaN so that I could perform a check to skip them later in the problem. Finally, I made a backup dataframe so that I could reference that information in case it was needed.

**Part B:** In the remainder of this problem, you will perform some hypothesis tests to examine whether these data suggest there are significant differences in the mean ratings of the different Oreo flavors. Pick a level of significance for these experiments, and explain how you decided to use that significance level. "Because we used it a lot in class" is ***not*** a good reason.

**ANSWER PART B**

Determining if there are significant differences in the mean ratings of Oreos is serious business! I am using a 99% level of significance in order to obtain an accurate assesment of the data.

**Part C:** Perform an **Analysis of Variance hypothesis test** in order to determine if there is evidence that there is _some_ difference among the mean ratings given to these 12 Oreo flavors. Clearly state your null and alternative hypothesis, and use the significance level identified in **Part B**. You must show **all** calculations **by hand** (and may of course use Python as a calculator, and to compute values from a distribution using the appropriate percent-point-function (ppf) or cumulative distribution function (cdf)).

In addition to showing the code for your calculations, make comments **in Markdown** explaining what you are doing.

**HYPOTHESIS PART C**

$H_0: \mu_1 = \mu_2 = ... = \mu_i$ There is not a significant difference between the mean ratings.

$H_1: \mu_i \ne \mu_j$ for some pair $i,j$ There is a significant difference between the mean ratings.

In [78]:
grandMean = dfOreo['average_across_raters'].mean()
SSB = 0
SSW = 0
SSBdf = 11
SSWdf = 42
for i in range (12):
    SSB = SSB + 12*((dfOreo['average_across_raters'][i]-grandMean)**2) 

for i in range (12):
    mean = dfOreo.iloc[i,5]
    for j in range (5):
        if (np.isnan(dfOreo.iloc[i,j]) == False):
            SSW = SSW + ((dfOreo.iloc[i,j]-mean)**2)

SST = SSB+SSW
F = (SSB/SSBdf)/(SSW/SSWdf)
rej_reg = stats.f.ppf(.01,dfn=11,dfd=42)
p = 1-(stats.f.cdf(F,11,42))
print(rej_reg,p)



#print(dfOreo.iloc[0,5])       

0.2598909843303179 0.0008345423550680797


**ANSWER PART C**

1. Our first step will be to find the grand mean of all of the groups (flavors)

2. Next I will calculate the SSB and SSW values:
    
    SSB is equal to the number groups multiplied by each row mean minus the overall grand mean squared for each row all added together. It looks like this:
    $SSB = ([12(2.5-3.52)^2] + [12(2.75-3.52)^2] + [12(3.25-3.52)^2] + [12(3.4-3.52)^2] + [12(3.4-3.52)^2] + [12(3.4-3.52)^2] + [12(3.6-3.52)^2] + [12(3.7-3.52)^2] + [12(3.8125-3.52)^2] + [12(4-3.52)^2] + [12(4.05-3.52)^2] + [12(4.375-3.52)^2]) = 37.393$
    
    To calculate the SSW for all 54 values I subtract the mean for the given row from each score in that row, square that value, and add it to a running total of SSW values. It looks like this:
    $SSW = ( 1.5 - 2.5 ) ^2 +( 4.0 - 2.5 ) ^2 +( 2.0 - 2.5 ) ^2 +( 2.5 - 2.75 ) ^2 +( 3.0 - 2.75 ) ^2 +( 4.0 - 2.75 ) ^2 +( 1.5 - 2.75 ) ^2 +( 4.5 - 3.25 ) ^2 +( 3.5 - 3.25 ) ^2 +( 1.5 - 3.25 ) ^2 +( 3.5 - 3.25 ) ^2 +( 3.5 - 3.4 ) ^2 +( 3.0 - 3.4 ) ^2 +( 3.5 - 3.4 ) ^2 +( 3.0 - 3.4 ) ^2 +( 4.0 - 3.4 ) ^2 +( 2.5 - 3.4 ) ^2 +( 5.0 - 3.4 ) ^2 +( 2.5 - 3.4 ) ^2 +( 4.0 - 3.4 ) ^2 +( 3.0 - 3.4 ) ^2 +( 3.5 - 3.4 ) ^2 +( 4.0 - 3.4 ) ^2 +( 3.0 - 3.4 ) ^2 +( 3.0 - 3.4 ) ^2 +( 3.5 - 3.4 ) ^2 +( 4.0 - 3.6 ) ^2 +( 3.0 - 3.6 ) ^2 +( 4.5 - 3.6 ) ^2 +( 4.0 - 3.6 ) ^2 +( 2.5 - 3.6 ) ^2 +( 2.5 - 3.7 ) ^2 +( 1.0 - 3.7 ) ^2 +( 5.0 - 3.7 ) ^2 +( 5.0 - 3.7 ) ^2 +( 5.0 - 3.7 ) ^2 +( 3.0 - 3.8125 ) ^2 +( 4.0 - 3.8125 ) ^2 +( 4.0 - 3.8125 ) ^2 +( 4.25 - 3.8125 ) ^2 +( 3.5 - 4.0 ) ^2 +( 4.5 - 4.0 ) ^2 +( 3.0 - 4.0 ) ^2 +( 4.5 - 4.0 ) ^2 +( 4.5 - 4.0 ) ^2 +( 3.5 - 4.05 ) ^2 +( 4.0 - 4.05 ) ^2 +( 4.5 - 4.05 ) ^2 +( 4.0 - 4.05 ) ^2 +( 4.25 - 4.05 ) ^2 +( 4.5 - 4.375 ) ^2 +( 4.5 - 4.375 ) ^2 +( 4.5 - 4.375 ) ^2 +( 4.0 - 4.375 ) ^2 = 37.7593745$


3. Now we can calculate our SST which is SSB+SSW and equals = 75.152343745

4. Next we will calculate our degrees of freedom. $SSB_{DF} = I-1 = 12-1 = 11$ and $SSW_{DF} = N-I = 54-12 = 42$

5. Now we can calculate our test statistic F. $F=\frac{\tfrac{SSB}{SSB_{DF}}}{\tfrac{SSW}{SSW_{DF}}} =\frac{\tfrac{37.393}{11}}{\tfrac{37.7593745}{42}} = 3.781131266317065$

6. Finally we calculate the p-value $p = 1-stats.f.cdf(F,11,42) = .000835$. Because $.000835 < .01$ We reject the null hypothesis and say that there IS a significant difference between the mean ratings of the different flavors of Oreos. 

**Part D:** Perform a **t hypothesis test** to determine if there is evidence supporting the claim that cinnamon bun Oreos have a higher mean rating than PB&J Oreos. Use the significance level you identified in **Part B**. Clearly state your null and alternative hypotheses, your conclusions, and show all work. Again, you may not use any canned t-test function.

**HYPOTHESIS PART D**

$H_0: \mu_1 > \mu_2$ Cinnamon Bun Oreos have a higher mean rating than PB&J Oreos

$H_1: \mu_i \le \mu_j$ Cinnamon Bun Oreos do not have a higher mean rating than PB&J Oreos

In [79]:
pbjMean = dfOreo.iloc[0,5]
cbMean = dfOreo.iloc[11,5]
pbjSD = dfOreo.iloc[0,0:3].std()
cbSD = dfOreo.iloc[11,1:5].std()
print(pbjMean,pbjSD,cbMean,cbSD)
t = (cbMean - pbjMean)/(np.sqrt(((cbSD**2)/4)+((pbjSD**2)/3)))
df = 2
p = stats.t.cdf(t,df=2)
print(t,p)


2.5 1.3228756555322954 4.375 0.25
2.4227185592617446 0.9318148774623103


**ANSWER PART D**

Peanut Butter and Jelly Mean = 2.5
Peanut Butter and Jelly SD = 1.3229
Cinnamon Bun Mean = 4.375
Cinnamon Bun SD = .25

$t = \frac{4.375-2.5}{\sqrt{\tfrac{.25^2}{4}+\tfrac{1.3229^2}{3}}} = 2.42271856$

Using a conservative degrees of freedom value of 2 we get:

$p = stats.t.cdf(2.42271856, df=2) = .93181488$

Because $.93 > .01$  we fail to reject the Null Hypothesis and can say that Cinnamon Bun Oreos have a higher mean rating than PB&J Oreos

**Part E:** Do your results from Parts C and D agree with one another? If they agree, explain how they are in agreement in _words_. If they do not agree, explain why you think they do not agree.

_Hint: if they do not agree, consider carefully what assumptions we make in ANOVA and the t-test._

**ANSWER PART E**

They do agree! In part C, the null hypothesis stated that all of the means of the samples were the same. We found that we needed to reject the null hypothesis thus concluded that there is a difference between the sameple means. In part D the hypothesis was that there was a difference between the means. The p value was much larger than the alpha value so we failed to reject the null hypothesis and again determined that there was a difference between the means. 

<br>

---
<a id='p2'></a>
[Back to top](#top)

### [30 points] Problem 2: Multiple Linear Regression to Explain House Hauntings

<img src="https://s-media-cache-ak0.pinimg.com/originals/09/72/01/09720128cff5de4d4af038cd3fcf7f69.jpg" style="width: 300px;"/>

In an effort to control the skyrocketing prices of real estate in the Colorado Front Range, Governor Polis implemented a cutting edge new intervention. This new program oversaw the introduction of ghosts back into their natural ecosystem, after the ghost population seriously dwindled in recent decades due to overhaunting. However, an unfortunate miscalculation has led to haunted houses becoming a very serious problem in Colorado. Modern problems require modern solutions, so Governor Polis has hired you and the famous hedgehog data scientist/part-time ghostbuster Amy to determine what features of a house may be used to best predict a `haunted` score, related to the probability that a house with the given features is haunted (higher $\leftrightarrow$ more likely to be haunted).

You decide to use multiple linear regression to understand and predict what factors lead to increased haunted house hazard. You collected a data set from Haunted Zillow, the lesser-known database of haunted house prices and attributes. The data cover a variety of potential features, and you'll find this data in the file `houses.csv`. 

**Response**: 

- $\texttt{haunted}$: a haunting score, related to the probability that a house with the given features is haunted (higher $\leftrightarrow$ more likely to be haunted)

**Features**: 

- $\texttt{age}$: age of the house, in years
- $\texttt{area}$: square footage of interior of house
- $\texttt{bathrooms}$: number of bathrooms
- $\texttt{distance metro}$: distance to the nearest major metropolitan area (in miles)
- $\texttt{distance cemetery}$: distance to the nearest cemetery (in miles)
- $\texttt{cats}$: the number of cats within a one-block radius of the house
- $\texttt{howls}$: the number of wolf howls heard on an average night in the house's neighborhood
- $\texttt{clouds}$: what percentage of the sky was covered by clouds (fraction, 0-1)
- $\texttt{precipitation}$: amount of precipitation in the past 72 hours (inches)
- $\texttt{misery index}$: an economic indicator for how miserable the average United States citizen is, based on the unemployment rate and the inflation rate. More [here](https://www.stuffyoushouldknow.com/podcasts/whats-the-misery-index.htm) and [here](https://en.wikipedia.org/wiki/Misery_index_(economics)). Higher values correspond to more miserable citizens.
- $\texttt{ice cream sold}$: the number of units of ice cream sold at the farmer's market the week the house was most recently sold

**Part A**: Read the data from `houses.csv` into a Pandas DataFrame.  Note that since we will be doing a multiple linear regression we will need all of the features, so you should drop any row in the DataFrame that is missing data. 

In [80]:
dfHouses = pd.read_csv("D:\Valentine\CU\CSCI-3022\Practicum_2\houses.csv")
dfHouses = dfHouses.dropna()

dfHouses.head(1000)


age  area  bathrooms  distance metro  distance cemetery  cats  howls  \
0    65.06  2041        1.0             7.1              10.01     7      3   
1   141.48  1564        0.0             7.4               4.07     5      5   
2     7.58  1637        3.0             7.0               3.36     2      0   
3    51.47  2021        2.0             7.9               3.43     6      8   
4   259.44  1642        1.0             7.5               3.19     4      1   
5   287.93  2676        2.0             8.1               4.57     7      6   
6    86.92  2494        0.0             7.1               4.69     4      4   
7   184.49  1157        3.0             6.8               3.13     6      4   
8    42.52  2520        1.0             6.1               3.42     3      3   
9   150.29  1790        1.0             5.6               4.44     4      9   
10  143.29  2389        2.0             8.1               5.49     2      2   
11   41.97  2019        4.0             7.4               6.04     4      7   
12  262.91  1985        2.0             5.5               6.88     7      7   
13    7.31  2717        3.0             5.9               5.41     3      5   
15  296.44  1943        2.0             6.3               3.65     3      4   
16  150.24  2002        5.0             6.6               3.45     4      7   
17  183.91  1812        5.0             7.5               3.43     5      9   
18  131.00  1462        3.0             7.3              10.56     4      7   
19  244.76  2268        2.0             7.4               6.23     6     10   
20  257.84  3010        3.0             6.6               3.26     3      5   
21  209.18  2132        3.0             6.1               4.19     4      2   
22   47.93  1204        3.0             8.4               3.08     7      5   
23  161.61  2377        1.0             8.3               6.80     1      8   
24  272.04  1914        3.0             6.7              15.67     6      7   
25  260.65  2013        2.0             6.0               3.20     3      7   
27   18.11  2705        4.0             8.4               3.49     6      9   
28   57.38  3239        4.0             5.5               7.49     9      5   
29   58.32  1708        2.0             7.6               4.02     4      3   
30   68.41  2114        3.0             5.5               3.03     4      4   
31   28.96  2576        4.0             6.9               3.63     5      4   
..     ...   ...        ...             ...                ...   ...    ...   
41  108.81  2290        2.0             6.4               4.18     8      8   
42   10.73  2132        2.0             7.9               3.33     5      9   
43  224.76  1691        1.0             5.6               3.36    12      4   
44   75.47  1668        0.0             7.0               3.52     3      3   
45   43.20  1969        3.0             8.2               4.22     4      5   
46  257.67  1915        3.0             7.8              73.38     7      6   
47   36.35   948        5.0             8.0               5.33     6      8   
48   36.56  1375        2.0             7.3               7.00     7      3   
49  253.65  2196        2.0             7.6               4.64    17      8   
50  219.30  1891        3.0             6.3               3.52     3      7   
51  200.38  2171        1.0             6.9               7.21    12      2   
52  102.13  2126        2.0             7.6               3.30     9      5   
53   49.38  1824        2.0             7.0               4.59     0      5   
54  135.57  2558        3.0             8.4               3.72     5      3   
55  161.73  2122        0.0             6.9               3.29     9     12   
56  130.61  1441        1.0             6.6               6.89     2      2   
57   19.50  1906        3.0             5.8               4.53     9      4   
58  250.40  2232        3.0             6.4               5.26     4      7   
59  264.88  1758        4.0             6.9              

**Part B**: Perform the appropriate statistical test at the $\alpha = 0.01$ significance level to determine if _at least one_ of the features is related to the the response $y$.  Clearly describe your methodology and show all computations in Python. 

$H_0: B_1 = B_2 = B_3 = B_4 = B_5 = B_6 = B_7 = B_8 = B_9 = B_{10} = B_{11} = 0$

$H_1: B_k \ne 0$ for at least one value of k in 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11

Using the code that we went over in notebook 23 I am creating an OLS (Ordinary Least Squares) model. For the response I am using the "haunted" column and the features will be all of the other columns from the data frame. A P>|t| value less than our .01 significance level will indicate that there is a relationship between the feature and the response.

In [81]:
y, X = dfHouses["haunted"], dfHouses.iloc[:,:11] 
X = sm.add_constant(X)
model = sm.OLS(y, X).fit() 
slopesFull = model.params
model.summary()
#print(model.fstatistic)

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                haunted   R-squared:                       0.948
Model:                            OLS   Adj. R-squared:                  0.938
Method:                 Least Squares   F-statistic:                     94.60
Date:                Wed, 01 May 2019   Prob (F-statistic):           1.97e-32
Time:                        14:30:47   Log-Likelihood:                 13.575
No. Observations:                  69   AIC:                            -3.151
Df Residuals:                      57   BIC:                             23.66
Df Model:                          11                                         
Covariance Type:            nonrobust                                         
=====================================================================================
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
const                 0.3480      0.323      1.078      0.285      -0.298       0.994
age                   0.0017      0.000      5.859      0.000       0.001       0.002
area                5.83e-05   6.16e-05      0.946      0.348   -6.51e-05       0.000
bathrooms            -0.0411      0.023     -1.806      0.076      -0.087       0.004
distance metro       -0.0142      0.032     -0.443      0.660      -0.079       0.050
distance cemetery    -0.1000      0.003    -30.867      0.000      -0.106      -0.093
cats                  0.0316      0.010      3.258      0.002       0.012       0.051
howls                 0.0056      0.011      0.494      0.623      -0.017       0.029
clouds                0.0902      0.123      0.734      0.466      -0.156       0.336
precipitation         0.0068      0.067      0.102      0.919      -0.128       0.141
misery index         -0.0082      0.006     -1.456      0.151      -0.020       0.003
ice cream sold       -0.0006      0.001     -1.274      0.208      -0.002       0.000
==============================================================================
Omnibus:                        1.629   Durbin-Watson:                   2.191
Prob(Omnibus):                  0.443   Jarque-Bera (JB):                1.480
Skew:                           0.230   Prob(JB):                        0.477
Kurtosis:                       2.450   Cond. No.                     2.55e+04
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 2.55e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

**ANSWER PART B**

We can see from the OLS model that there is a relationship between the number of cats and the response, the age of the house and the response, and the distance to the cemetery and the response. Thus, we will reject the Null Hypothesis and say that there is at least one feature that relates to the response.

**Part C**: Write a function `forward_select(df, resp_str, maxk)` that takes in the DataFrame, the name of the column corresponding to the response, and the maximum number of desired features, and returns a list of feature names corresponding to the `maxk` most important features via forward selection.  At each stage in forward selection you should add the feature whose inclusion in the model would result in the lowest sum of squared errors $(SSE)$. Use your function to determine the best $k=5$ features to include in the model. Clearly indicate which feature was added in each stage. 

**Note**: The point of this exercise is to see if you can implement **foward_select** yourself.  You may of course use canned routines like statmodels OLS, but you may not call any Python method that explicitly performs forward selection.

In [82]:
def forward_select(df, resp_str, maxk):
    slopes = []
    intercepts = []
    SSE = 0
    tempSSE = 0
    position = 0
    positions = []
    bestFeatures = []


    variables = ["age", "area", "bathrooms", "distance metro", "distance cemetery", "cats", "howls", "clouds", "precipitation", "misery index", "ice cream sold"]
    for v in variables: 
        #print("SLR for {} vs hauntedness".format(v))
        #print("----------------------")
        bhat, ahat, rval, pval, stderr = stats.linregress(df[v], df[resp_str])
        slopes.append(bhat)
        intercepts.append(ahat)
        #print("intercept = {:.4f}".format(ahat))
        #print("slope = {:.4f}".format(bhat))
        #print("p-value = {}".format(pval))
        #print("\n")


    # Baseline SSE
    for i in range (69):
        SSE = SSE + ((df.iloc[i,11] - 1)**2)

    # Adding the best feature using SSE simple linear regression
    for i in range (11):
        slope = slopes[i]
        intercept = intercepts[i]
        for j in range (69):
            tempSSE = tempSSE + ((df.iloc[j,11] - (intercept + (slope*df.iloc[j,i])))**2)
        if (tempSSE < SSE):
            SSE = tempSSE
            position = i
        tempSSE = 0
    print("Added {} as a feature" .format(variables[position]))
    positions.append(position)
    bestFeatures.append(variables[position])
    
    bigX = df.iloc[:,positions[0]:positions[0]+1]
    
    # Adding the remaining features using R squred MLR
    for i in range (maxk-1):
        r = 0
        for j in range (11):
            valid = True
            for k in range(len(positions)):
                if (j == positions[k]):
                    valid = False
            if (valid == True):
                tempR = 0
                y = df[resp_str]
                X = bigX
                X = X.join(df.iloc[:,j])
                X = sm.add_constant(X)
                model = sm.OLS(y, X).fit()
                tempR = model.rsquared
                if (tempR > r):
                    r = tempR
                    position = j
        bigX = bigX.join(df.iloc[:,position])
        print("Added {} as a feature" .format(variables[position]))
        bestFeatures.append(variables[position])
        positions.append(position)
    
    return(bestFeatures)

print(forward_select(dfHouses,"haunted",5))

Added distance cemetery as a feature
Added age as a feature
Added cats as a feature
Added bathrooms as a feature
Added misery index as a feature
['distance cemetery', 'age', 'cats', 'bathrooms', 'misery index']


In [83]:
y, X = dfHouses["haunted"], dfHouses.iloc[:,[0,4,5,2,9]] 
X = sm.add_constant(X)
model = sm.OLS(y, X).fit() 
slopesRed = model.params
model.summary()


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                haunted   R-squared:                       0.946
Model:                            OLS   Adj. R-squared:                  0.941
Method:                 Least Squares   F-statistic:                     218.7
Date:                Wed, 01 May 2019   Prob (F-statistic):           2.00e-38
Time:                        14:30:51   Log-Likelihood:                 11.932
No. Observations:                  69   AIC:                            -11.86
Df Residuals:                      63   BIC:                             1.541
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
=====================================================================================
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
const                 0.3456      0.107      3.220      0.002       0.131       0.560
age                   0.0018      0.000      6.506      0.000       0.001       0.002
distance cemetery    -0.1005      0.003    -32.801      0.000      -0.107      -0.094
cats                  0.0327      0.009      3.545      0.001       0.014       0.051
bathrooms            -0.0382      0.022     -1.747      0.085      -0.082       0.005
misery index         -0.0087      0.005     -1.592      0.116      -0.020       0.002
==============================================================================
Omnibus:                        1.395   Durbin-Watson:                   2.178
Prob(Omnibus):                  0.498   Jarque-Bera (JB):                1.434
Skew:                           0.289   Prob(JB):                        0.488
Kurtosis:                       2.595   Cond. No.                         739.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

**Part D**: Write down the multiple linear regression model, including estimated parameters, obtained by your forward selection process. 

**ANSWER PART D**

$haunted = 0.3456 - (0.1005*Distance Cemetery) + (0.0018*Age)  + (0.0327*Cats) - (0.0382*Bathrooms) - (0.0087*Misery Index)$ 

**Part E**: Perform the appropriate statistical test at the $\alpha = 0.05$ significance level to determine whether there is a statistically significant difference between the full model with all features and the reduced model obtained by forward selection in **Part D**. 

**ANSWER PART E**

$H_0: B_1 = B_3 = B_5 = B_6 = B_7 = B_8 = B_{10} = 0$

$H_1: B_k \ne 0$ for at least one value of k in 1, 3, 5, 6, 7, 8, 10

In [84]:
SSEF = 0
SSER = 0
for i in range (69):
    SSEF = SSEF + (dfHouses.iloc[i,11] - (slopesFull[0] + (slopesFull[1]*dfHouses.iloc[i,0]) + (slopesFull[2]*dfHouses.iloc[i,1]) + (slopesFull[3]*dfHouses.iloc[i,2]) + (slopesFull[4]*dfHouses.iloc[i,3]) + (slopesFull[5]*dfHouses.iloc[i,4]) + (slopesFull[6]*dfHouses.iloc[i,5]) + (slopesFull[7]*dfHouses.iloc[i,6]) + (slopesFull[8]*dfHouses.iloc[i,7]) + (slopesFull[9]*dfHouses.iloc[i,8]) + (slopesFull[10]*dfHouses.iloc[i,9]) + (slopesFull[11]*dfHouses.iloc[i,10])))**2
for i in range (69):
    SSER = SSER + (dfHouses.iloc[i,11] - (slopesRed[0] + (slopesRed[1]*dfHouses.iloc[i,0])  + (slopesRed[2]*dfHouses.iloc[i,4]) + (slopesRed[3]*dfHouses.iloc[i,5]) + (slopesRed[4]*dfHouses.iloc[i,2]) + (slopesRed[5]*dfHouses.iloc[i,9])))**2

F = ((SSER - SSEF)/(11-5))/((SSEF)/(69-11-1))
print(F)

0.46348704337579594


Because $0.463487 > \alpha$ with $\alpha = .05$ we fail to reject the Null Hypothesis and can say that there is no statistically significant difference between the full model and the reduced model at the $\alpha = 0.05$ significance level

**Part F**: Based on your conclusions in **Part E**, use the _better_ of the two models to predict the sharknado hazard when the following features are observed: 

- $\texttt{age}$: 100 years
- $\texttt{area}$: 2200 square feet
- $\texttt{bathrooms}$: 3 bathrooms
- $\texttt{distance metro}$: 25 miles
- $\texttt{distance cemetery}$: 10 miles
- $\texttt{cats}$: 4 cats
- $\texttt{howls}$: 5 wolf howls/night
- $\texttt{clouds}$: 0.65 cloud cover
- $\texttt{precipitation}$: 0 inches
- $\texttt{misery index}$: 10
- $\texttt{ice cream sold}$: 125

**ANSWER PART F**

Because the adjusted Rscore from the partial model is higher than the adjusted Rscore of the full model, we can say it is the better model for making a prediction.

we recall our formula from Part D

$Sharknado? = 0.3456 - (0.1005*Distance Cemetery) + (0.0018*Age)  + (0.0327*Cats) - (0.0382*Bathrooms) - (0.0087*Misery Index)$ 

In [85]:
haunted = (.3456 - (.1005 * 10) + (.0018 * 100) + (.0327 * 4) - (.0382 * 3) - (.0087 * 10))
print(haunted)

-0.5502000000000001


This yields -0.5502... meaning that the likelyhood of a localized sharknado event is relatively low on the Front Range compared to our dataset, i.e. most of the values in our original dataframe.

Also, I think you meant haunted not sharknado but I'm rolling with it.

**Part G:** Governor Polis dabbles a bit in the art of data science, as well as the science of data art. He tells you that the response (`haunted` score) that you and Amy predicted is actually the natural logarithm of the _odds_ that a house with the given features is haunted, where if $p$ is the probability that a house is haunted, then the odds are given by $$\text{odds} = \dfrac{p}{1-p}$$

What transformation must you make in order to turn your multiple regression model into a logistic regression model, to classify a house a haunted or not? Perform this simple transformation, then use a decision threshold of 0.5 to classify the house from **Part F** as haunted or not haunted. No new models should be fit here; use the same model that you used in Part F.

**ANSWER PART G**

In order to transform to a logistic regression model we use the following math

In [86]:
odds = np.exp(-.5502)
p = (.568/1.568)
print("the probability of a house being haunted is {} and the odds are {}" .format(p,odds))

the probability of a house being haunted is 0.36224489795918363 and the odds are 0.5768344319566375


Thus, using a 0.5 decision threshold we can say that the house is not haunted... and probably won't be hit by a sharknado either :)

<br>

---
[Back to top](#top)
<a id='bot'></a>